runtime/compiler/trj9/z/codegen/J9S390SystemLinkage.cpp

```diff
@@ -395,24 +395,35 @@ TR::J9S390zOSSystemLinkage::generateInstructionsForCall(TR::Node * callNode, TR:
          }
       }
 
-   gcPoint = generateRRInstruction(codeGen, TR::InstOpCode::BASR, callNode, systemReturnAddressRegister, systemEntryPointRegister, deps);
+   // NOP padding is needed because return from C function will skip the XPLink eyecatcher and
+   // always return to a point that's 4 bytes after the return address.
+   // The 4-byte NOP is there to ensure entry to valid instruction.
+   // The BASR and NOP must stick together and can't have reverse spills in the middle.
+   // Hence, splitting the dependencies to avoid spills.
+   TR::RegisterDependencyConditions* callPreDeps = new (self()->trHeapMemory()) TR::RegisterDependencyConditions(deps->getPreConditions(), NULL, deps->getAddCursorForPre(), 0, codeGen);
+   TR::RegisterDependencyConditions* callPostDeps = new (self()->trHeapMemory()) TR::RegisterDependencyConditions(NULL, deps->getPostConditions(), 0, deps->getAddCursorForPost(), codeGen);
+
+   gcPoint = generateRRInstruction(codeGen, TR::InstOpCode::BASR, callNode, systemReturnAddressRegister, systemEntryPointRegister, callPreDeps);
    if (isJNIGCPoint)
          gcPoint->setNeedsGCMap(0x00000000);
 
    callType = TR_XPLinkCallType_BASR;
 
    generateS390LabelInstruction(codeGen, TR::InstOpCode::LABEL, callNode, returnFromJNICallLabel);
 
+   TR::Instruction * cursor = NULL;
    if (TR::Compiler->target.is64Bit())
       {
       //In 64 bit XPLINK, the caller returns at RetAddr+2, so add 2 byte nop
-      TR::Instruction * cursor = new (trHeapMemory()) TR::S390NOPInstruction(TR::InstOpCode::NOP, 2, callNode, codeGen);
+      cursor = new (trHeapMemory()) TR::S390NOPInstruction(TR::InstOpCode::NOP, 2, callNode, codeGen);
```

Comment:
Can you use the `OMR::Z::CodeGenerator::insertPad` API here instead?

Reply:
OK. I'll try insertPad API.

Reply:
I refactored and simplified the BASR padding function in system linkage class to use `insertPad()` API. I'm currently running tests.